remove stubbed unbound modules, move rest of DOM classes to DOM.js (fixes #175)

ngokevin opened the pull request:
- When bindings are set, store in a Node module to be referenced at runtime by classes, versus using a big closure.
- Removed stubs for `ImageData`, `ImageBitmap`, `Path2D`, `CanvasGradient`, `CanvasRenderingContext2D`, `WebGLRenderingContext`, `MicrophoneMediaStream`, `HTMLImageElement`, `HTMLAudioElement`, `HTMLVideoElement`.
- Moved definitions of `HTMLImageElement`, `HTMLAudioElement`, and `HTMLVideoElement` to DOM module.
- Basic tests for img, audio, audio.

modulesio commented on src/DOM.js:
```diff
@@ -10,6 +10,7 @@ const util = require('util');
 const {defaultCanvasSize} = require('./constants');
 const {Event, EventTarget, MouseEvent} = require('./Event');
 const GlobalContext = require('./GlobalContext');
+const bindings = require('../native-bindings');
```
Per Slack, probably don't do this. Native bindings is somewhat sensitive to when you require it, because of `Worker` threading.

https://github.com/webmixedreality/exokit/blob/0475da06f10938bdc2525ea3d8056ab6df3e2f4a/core.js#L1695

But that might be fixed when we move to Node's new `worker_threads`, which we should when it's stable.

modulesio replied:
Then again, https://nodejs.org/api/worker_threads.html:

> Native addons are not supported yet.

Probably for similar reasons.

ngokevin commented:
Updated. I set the bindings in a Node module to be referenced at runtime by classes, versus waiting to define classes at runtime and using a big closure for classes to have reference to bindings.

Added some tests.

modulesio commented on tests/unit/dom/MediaElement.test.js:
```diff
@@ -0,0 +1,43 @@
+  describe('<video>', () => {
+    it('can set src', () => {
+      el = document.createElement('video');
+      el.setAttribute('src', videoUri);
```
Should probably wait for `onload`/`onerror`. Otherwise a whole class of bugs potentially fires after the test has passed.

modulesio replied:
Sorry, for video/audio I think it's `oncanplay`.

modulesio commented on core.js:
```diff
@@ -1685,11 +1592,24 @@ exokit.setArgs = newArgs => {
   const bindings = require(nativeBindingsModule);
 
+  // Set in binding module to be referenced from other modules.
+  for (let key in bindings) { BindingsModule[key] = bindings[key]; }
```
Nit: could be `const`.

ngokevin replied:
Can it? It's an iterator variable.

modulesio replied:
Pretty sure.

```
o = {lol: 'zol', root: 'toot'};
for (const k in o) {
  console.log('key', k);
}
```

The exception is `for-i` where you assign to the loop variable yourself.

modulesio commented on core.js:
```diff
@@ -1685,11 +1592,24 @@ exokit.setArgs = newArgs => {
 
+  // Set in binding module to be referenced from other modules.
+  for (let key in bindings) { BindingsModule[key] = bindings[key]; }
+  console.log(BindingsModule);
```
Probably should keep it clean.

modulesio commented:
Design looks ✔️

ngokevin commented:
Updated the tests with assertion that the media gets loaded. I run a little Express server to serve files locally that will get fetched live by Exokit.

I fixed Video element not having canplay/canplaythrough, and I refactored all those handlers up to the MediaElement.

modulesio commented:
> I run a little Express server to serve files locally that will get fetched live by Exokit.

I think `file://` might work too. But it's good to test network since that's the main one.

modulesio commented on tests/unit/dom/SrcableElement.test.js:
```diff
@@ -0,0 +1,48 @@
+const exokit = require('../../../index');
+
+const audioData = fs.readFileSync(path.resolve(__dirname, '../data/test.ogg'), 'base64');
+const audioUri = `data:audio/ogg;base64,${audioData}`;
```
This would be a good test, if it were used :).

I can do a follow-up with some more cases.
